cloudsync/smartsync.py

```diff
@@ -0,0 +1,186 @@
+        return self.state.changeset_len
+
+    @property
+    def changes(self):
+        return super().changes
+
+    def pre_sync(self, sync: SyncEntry) -> bool:  # pylint: disable=too-many-branches
+        finished = super().pre_sync(sync)
+        if not finished:
+            finished = not (sync in self.state.requestset or sync[REMOTE].otype == DIRECTORY)
+        return finished
+
+    def get_parent_conflicts(self, sync: SyncEntry, changed) -> List[SyncEntry]:
+        done = False
+        pcs = []
+        new_pc = sync
+        while not done:
+            pc = self._get_parent_conflict(new_pc, changed)
+            if pc and pc[REMOTE].path != new_pc[REMOTE].path:
+                pcs.append(pc)
+                new_pc = pc
+            else:
+                done = True
+        pcs.reverse()
+        return pcs
+
+
+class SmartSyncState(SyncState):
+    def __init__(self,
+                 providers: Tuple['Provider', 'Provider'],
+                 storage: Optional['Storage'] = None,
+                 tag: Optional[str] = None,
+                 shuffle: bool = False,
+                 prioritize: Callable[[int, str], int] = None):
+        self.requestset = set()
+        self.excludeset = set()
+        super().__init__(providers, storage, tag, shuffle, prioritize)
+
+    def smart_sync_ent(self, ent):
+        self.requestset.add(ent)
+        self.excludeset.discard(ent)
+
+    def smart_sync_path(self, remote_path) -> List[SyncEntry]:
+        # q: do we want to raise an exception if path isn't found?
+        # q: do we need to also sync path's parents all the way to the root?
+        #   a: probably, but it doesn't make sense to add them to the requestset, just sync them implicitly
+        #       or maybe it does make sense to add them to the requestset? that way they will get subscribed too
+        #       or maybe always sync all folders at all times, and only files are smartsynced
+        ents = self.lookup_path(REMOTE, remote_path)
+        for ent in ents:
+            self.requestset.add(ent)
+        return ents
+
+    def smart_unsync_ent(self, ent):
+        self.requestset.discard(ent)
+        self.excludeset.add(ent)
+
+    def smart_unsync_path(self, remote_path):
+        for ent in tuple(self.requestset):
+            if ent[REMOTE].path == remote_path:
+                self.state.smart_unsync_ent(ent)
+                break
+
+    def smart_listdir(self, local_path, remote_path):
+        prov = self.providers[REMOTE]
+        for ent, _ignored_rel_path in self.get_kids(remote_path, REMOTE):
+            remote_ent_path = ent[REMOTE].path
+            _ignored_remote_ent_parent, remote_ent_name = self.providers[REMOTE].split(ent[REMOTE].path)
+            if prov.paths_match(prov.dirname(remote_ent_path), remote_path):
+                ent_info = prov.info_oid(ent[REMOTE].oid)
+                retval = DirInfo(otype=ent[REMOTE].otype,
+                                 oid='',
```

Comment:
include oid

Reply:
Do we need to know if the file is in the exclusion list here?

Reply:
oid is included in the most recent commit. this should expose the ent if it is in the exclude list. The exclude list is only a list of files that are explicitly unsynced. The whole point of smart_listdir is to include unsynced files